### PR TITLE
Initial MCP server setup

### PR DESCRIPTION
**Pull request by MacondoExpress**
# Initial Neo4j MCP Project Setup

## Summary
This PR establishes the foundational structure for the Neo4j Model Context Protocol (MCP) server implementation. It includes a basic prototype of the `read-cypher` tool to demonstrate core functionality, but is not production-ready.

**Note**: This PR also serves to test the repository workflow and CI/CD processes on a non-empty project structure.

This is a **placeholder implementation** and lacks several production requirements.

### CODE REVIEW

File: internal/server/server.go [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+}
+
+// NewNeo4jMCPServer creates a new MCP server instance
+func NewNeo4jMCPServer() *Neo4jMCPServer {
+	mcpServer := server.NewMCPServer(
+		"neo4-mcp",
+		"0.0.1",
+	)
+
+	cfg := config.LoadConfig()
```
Comment (angrykoala) [inline internal/server/server.go]: Consider passing the cfg as a parameter rather than loading it here

File: internal/server/server.go [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+}
+
+// Define ReadCypher input schema
+type ReadCypherInput struct {
+	Query  string         `json:"query" jsonschema:"default=MATCH(n) RETURN n,description=The Cypher query to execute"`
+	Params map[string]any `json:"params" jsonschema:"description=Parameters to pass to the Cypher query"`
+}
+
+// RegisterTools registers all available MCP tools
+func (s *Neo4jMCPServer) RegisterTools() {
```
Comment (angrykoala) [inline internal/server/server.go:50]: This seems to be all dealing with private information, any reason for not doing this in the Start function?

File: internal/server/server.go [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+	if Params != nil {
+		fmt.Fprintf(os.Stderr, "cypher-parameters: %v\n", Params)
+	}
+	fmt.Fprintf(os.Stderr, "url: %s\n", s.config.URI)
+	fmt.Fprintf(os.Stderr, "database: %s\n", s.config.Database)
+
+	// Execute the Cypher query using the stored driver
+	response, err := s.executeReadQuery(ctx, Query, Params)
+	if err != nil {
+		return mcp.NewToolResultError(fmt.Sprintf("Error executing Cypher query: %v", err)), nil
```
Comment (keremgocen) [inline internal/server/server.go:100]: I think this line should return the err instead of wrapping it inside the NewToolResultError? or was that intentional?

File: internal/server/server.go [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+	ctx := context.Background()
+	return (*s.driver).Close(ctx)
+}
+
+// handleReadCypher handles the read-cypher tool requests
+func (s *Neo4jMCPServer) handleReadCypher(ctx context.Context, request mcp.CallToolRequest) (*mcp.CallToolResult, error) {
+	var args ReadCypherInput
+	// Bind arguments to the struct
+	if err := request.BindArguments(&args); err != nil {
+		return mcp.NewToolResultError(err.Error()), nil
```
Comment (keremgocen) [inline internal/server/server.go:84]: same here about err return being in reverse order, it feels more natural to return the error and bubble it up to be handled at the upper level in golang, I might need to check how github.com/mark3labs/mcp-go/mcp handles with errors as well.

File: internal/server/server.go [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+	var results []map[string]any
+	for _, record := range res.Records {
+		recordMap := make(map[string]any)
+		for i, key := range record.Keys {
+			recordMap[key] = record.Values[i]
+		}
+		results = append(results, recordMap)
+	}
+
+	formattedResponse, err := json.MarshalIndent(results, "", "  ")
```
Comment (keremgocen) [inline internal/server/server.go:125]: the formatting here feels a bit brittle, I'd move it to a separate function and add some tests against to make sure it behaves as we expect and modify it easily but the whole query execution logic is in progress I'm guessing right now. (then this function also becomes `excute` -> `parse` -> `handle error if any or return`)

File: internal/server/server.go (hunk 2 above)
Comment (MacondoExpress) [reply internal/server/server.go:50]: It will soon be changed to handle more logic, registering a different set of tools depending on the configuration/environment. Also, likely in the future, there will also be other register functions such as "registerPrompts", "registerResources", etc Depending on the future implementation, I will expose these "register" functions or just invoke them from the "Start" method.

File: internal/server/server.go (hunk 3 above)
Comment (MacondoExpress) [reply internal/server/server.go:100]: It was intentional, 
I wrapped any errors happening in the tool inside the "NewToolResultError" from the mark3labs method signature:
```
// NewToolResultError creates a new CallToolResult with an error message.
// Any errors that originate from the tool SHOULD be reported inside the result object.
```
From the spec:
https://modelcontextprotocol.io/specification/2025-06-18/server/tools#error-handling
I'm not 100% sure here, I feel is the "most right" given the nature of the tool, however, I think we should have a conversation/guidelines regarding the error handling, especially for benefit to external contributors! For instance, maybe in this case Cypher errors are handled correctly, but "network error" not. 
Thanks for raising it!

File: internal/server/server.go (hunk 4 above)
Comment (MacondoExpress) [reply internal/server/server.go:84]: same as above.

File: internal/server/server.go (hunk 5 above)
Comment (MacondoExpress) [reply internal/server/server.go:125]: Definitely,  it's already in the plan, I'll keep it out of the scope of this PR as this PR was mainly to have a starting point to test the repository ergonomics.
